File: hack/toys/load-controller/controller.go
package main

import (
	"context"

	"github.com/aperturerobotics/controllerbus/controller"
	"github.com/aperturerobotics/controllerbus/directive"
	"github.com/blang/semver"
	"github.com/sirupsen/logrus"
)

// ControllerID is the controller ID.
const ControllerID = "bifrost/toys/load-controller/controller/1"

// Version is the controller version.
var Version = semver.MustParse("0.0.1")

// ToyController is an example controller.
type ToyController struct {
	// le is the logger
	le *logrus.Entry
	// conf is the config
	conf *ToyControllerConfig
}

// NewToyController constructs a new toy controller.
func NewToyController(le *logrus.Entry, conf *ToyControllerConfig) (*ToyController, error) {
	return &ToyController{le: le, conf: conf}, nil
}

// GetControllerInfo returns information about the controller.
func (c *ToyController) GetControllerInfo() controller.Info {
	return controller.NewInfo(
		ControllerID,
		Version,
		"toy controller",
	)
}

// HandleDirective asks if the handler can resolve the directive.
func (c *ToyController) HandleDirective(di directive.Instance) (directive.Resolver, error) {
	// Pass for the example controller
	return nil, nil
}

// Execute executes the given controller.
// Returning nil ends execution.
// Returning an error triggers a retry with backoff.
func (c *ToyController) Execute(ctx context.Context) error {
	c.le.Debug("toy controller executed")
	<-ctx.Done()
	return nil
}

// SayHello says hello.
func (c *ToyController) SayHello() {
	c.le.Debugf("Hello %s!", c.conf.GetName())
}

// Close releases any resources used by the controller.
// Error indicates any issue encountered releasing.
func (c *ToyController) Close() error {
	return nil
}

// _ is a type assertion
var _ controller.Controller = ((*ToyController)(nil))

File: example/hello-world/controller.ts
import { Context } from '../../core/core'
import { Controller, Info, newInfo } from '../../controller/controller'
import { Directive, Instance, Resolver } from '../../directive/directive'
import { Logger } from '../../core/core'
import { ToyControllerConfig } from './controller.pb'

// ControllerID is the controller ID.
export const ControllerID = 'controllerbus/toys/load-controller/controller'

// Version is the controller version.
export const Version = '0.0.1'

/**
 * ToyController is an example controller.
 */
export class ToyController implements Controller {
  private readonly logger: Logger
  private readonly conf: ToyControllerConfig

  /**
   * Constructs a new toy controller.
   */
  constructor(logger: Logger, conf: ToyControllerConfig) {
    this.logger = logger
    this.conf = conf
  }

  /**
   * GetControllerInfo returns information about the controller.
   */
  getControllerInfo(): Info {
    return newInfo(ControllerID, Version, 'toy controller')
  }

  /**
   * HandleDirective asks if the handler can resolve the directive.
   */
  async handleDirective(
    ctx: Context,
    instance: Instance,
  ): Promise<Resolver[] | null> {
    // Pass for the example controller
    return null
  }

  /**
   * Execute executes the controller goroutine.
   * Returning null ends execution.
   * Returning an error triggers a retry with backoff.
   */
  async execute(ctx: Context): Promise<Error | null> {
    this.logger.debug('toy controller executed')
    await ctx.done()
    return null
  }

  /**
   * SayHello says hello.
   */
  sayHello(): void {
    this.logger.debugf('Hello %s!', this.conf.name)
  }

  /**
   * Close releases any resources used by the controller.
   */
  async close(): Promise<Error | null> {
    return null
  }
}
